octodns/provider/powerdns.py

```diff
@@ -165,23 +169,78 @@ def _data_for_SRV(self, rrset):
             'ttl': rrset['ttl']
         }
 
+    def detect_version(self):
+        # Only detect version once
+        if self.version_detected != '':
+            self.log.debug('detect_version: version %s allready detected',
+                           self.version_detected)
+            return self.version_detected
+
+        try:
+            self.log.debug('detect_version: getting version from server')
+            resp = self._get('')
+        except HTTPError as e:
+            if e.response.status_code == 401:
+                # Nicer error message for auth problems
+                raise Exception('PowerDNS unauthorized host={}'
+                                .format(self.host))
+            else:
+                raise
+
+        self.version_detected = resp.json()["version"]
+        self.log.debug('detect_version: got version %s from server',
+                       self.version_detected)
+        self.configure_for_version(self.version_detected)
+
+    def configure_for_version(self, version):
+        major, minor, patch = version.split('.', 2)
+        major, minor, patch = int(major), int(minor), int(patch)
+        self.log.debug('configure_for_version: configure for '
+                       'major: %s, minor: %s, patch: %s',
+                       major, minor, patch)
+
+        # Defaults for v4.0.0
+        self.soa_edit_api = "INCEPTION-INCREMENT"
+        self.check_status_not_found = False
+
+        if major == 4 and minor >= 2:
+            self.log.debug("configure_for_version: Version >= 4.2")
+            self.soa_edit_api = "INCEPTION-INCREMENT"
+            self.check_status_not_found = True
+
+        if major == 4 and minor >= 3:
+            self.log.debug("configure_for_version: Version >= 4.3")
+            self.soa_edit_api = "DEFAULT"
+            self.check_status_not_found = True
+
     def populate(self, zone, target=False, lenient=False):
         self.log.debug('populate: name=%s, target=%s, lenient=%s', zone.name,
                        target, lenient)
 
+        self.detect_version()
```

Comment:
Currently it's detecting the version here in populate, but as far as I can tell only needs the info when things are applying. It's minor, but it's appears to be an unneeded step/round trip and just something else to potentially fail. 

Going to look at making a quick refactor of this logic so that it only pulls the info when it's needed and also explicitly makes sure that it's pull before its needed. Will push that up in a bit.

Reply:
Half-wrong here. This check is needed when zones are not found during populate. It's not needed when all zones exists though so it's still good to only do it when required.